cranelift/codegen/src/machinst/mod.rs

```diff
@@ -269,6 +272,48 @@ pub trait MachInstLabelUse: Clone + Copy + Debug + Eq {
     fn from_reloc(reloc: Reloc, addend: Addend) -> Option<Self>;
 }
 
+/// Classification of call instruction types for granular analysis.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum CallType {
+    /// Not a call instruction.
+    None,
+    /// Regular call that returns to the caller.
+    Regular,
+    /// Tail call that doesn't return to the caller.
+    TailCall,
+}
+
+/// Enhanced function type classification for optimization purposes.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum FunctionType {
+    /// Function makes no calls at all.
+    Leaf,
+    /// Function only makes tail calls.
+    TailCallOnly,
+    /// Function has regular calls (with or without tail calls).
+    NonLeaf,
+}
+
+impl CallType {
+    /// Check if this is any kind of call (regular or tail).
+    pub fn is_call(self) -> bool {
+        matches!(self, CallType::Regular | CallType::TailCall)
+    }
+}
+
+impl FunctionType {
+    /// Check if this function type requires full frame setup.
+    #[allow(dead_code)]
+    pub fn requires_frame_setup(self) -> bool {
+        !matches!(self, FunctionType::Leaf)
+    }
+
+    /// Convert to boolean leaf classification.
+    pub fn is_leaf(self) -> bool {
```

Comment:
This appears to be used only in tests and in passing to `compute_frame_layout`, and the latter can just take the enum type directly; could we remove this method?